README.rst
:parent project: http://freshmeat.net/projects/http-replicator
:homepage: http://github.com/dotmpe/htcache 

**htcache** aims to be a versatile caching and rewriting HTTP and FTP proxy.
It is a fork of http-replicator 4.0 alpha 2. See CHANGELOG.

.. contents::

Status
------
Todo
 - (auto) remove descriptors after manual path delete.
 - use strict and other modes, adhere to RFC 2616:

    - calculate Age field [14.6]  
    - don't cache Authorization response [14.8]
    - Cache-Control [14.9]

 - rules.sort prefixes paths

Issues
 1. Writing to client fails randomly, probably dropped connection 
    (eg. cancelled mouseovers)
 2. Some date headers in the wild still fail to parse.
 3. HTML placeholder served for all connections (e.g. also for flash, images)
 4. There is a version with other cl-options, it uses stdlib asyncore
    check: 
 
     * http://web.archive.org/web/20070816213819/gertjan.freezope.org/replicator/http-replicator
     * http://web.archive.org/web/20071214200800/gertjan.freezope.org/replicator
  
 5. Embedded youtube does not work, but the site runs fine.

Unittests
 No known failures.

Installation
------------
Start as any Python script, or:

- cp/link htcache into ``/usr/bin`` 
- cp/link ``init.sh`` into ``/dev/init.d/``, modify htcache flags as needed.
  Make sure paths in init.sh and Params.py are accessible.
- add line ``/etc/init.d/htcache start`` to ``/etc/local`` for 
  on-startup initialization.

See http://www.debian-administration.org/articles/28 for Debian specifics.

Also create files in /etc/htcache:

* rules.drop
* rules.nocache 
* rules.sort

Overview
--------
htcache client/server flow::

   .                         htcache 
                             _______
                                o <------------*request---  client
                                |
                                |---blocked response(1)--->
                                |---static response(7)---->
   server <------------normal---|
          <------*conditional---' 
           --*normal----------> o
                                |--*normal----------------> 
                                `--*nocache response(4)---> 
           ---not modified----> o--*cached response------->       
           ---error-----------> o---direct response------->       

   * indicates wether there may be partial entity-content transfer


Normally a request creates a new cache location and descriptor, static 
responses are always served from cache and conditional requests may be.

Beside these messages, also note the following special cases of request 
and response messages.

1. blocked response                                  (rules.drop)
4. blind response (uncached content)                 (rules.nocache)

See the section `Rule Syntax`_ for the exact syntax.


Configuration
~~~~~~~~~~~~~
There is no separate configuration file, see Params.py and init.sh for 
option arguments to the program, and for their default settings. Other settings
are given in the rewrite and rules files described above.

The programs options are divided in three parts, the first group affects 
the proxy server, which is the default action.

To manage the cached resources and their descriptors, additional
query and maintenance options are provided. Note that maintenance may need
exclusive write access to the cache and descriptor backends, meaning don't run
with active proxy.

Cache backends listing
~~~~~~~~~~~~~~~~~~~~~~
htcache uses a file-based Cache which may produce a file-tree similar to 
that of ``wget -r`` (except if ``--nodir`` or ``--archive`` is in effect). 
This can create problems with long filenames and the characters that appear 
in the various URL parts.

Additional backends address this. (default: Cache.File, ``--cache TYPE``)

- caches.FileTreeQ - encodes each query argument into a separate directory,
  the first argument being prefixed with '?'. FIXME: does not solve anything?
- caches.FileTreeQH - Converts query into a hashsum. This one makes a bit more
  sense because queries are not hierarchical. The hashsum is encoded to a
  directory, the name prefixed with '#'.
- caches.PartialMD5 - only encodes the excess part of the filename, the limit 
  being hardcoded to 256 characters.
- caches.FileTree - combines above three methods. 
- caches.RefHash - simply encodes full URI into MD5 hex-digest and use as
  filename.

The storage location is futher affected by ``--archive`` and ``--nodir``.

Regular archival of a resources is possible by prefixing a formatted date to
the path. Ie. '%Y/%M/%d' would store a copy and maintain updates of a 
resource for every day. Prefixing a timestamp would probably store a new copy 
for each request.

``--archive`` results in lots of redundant data. It also makes static, offline
proxy operation on the resulting filesystem tree impossible. 

The nodir parameter accepts a replacement for the directory separator and
stores the path in a single filename. This may affect FileTreeQ.

Descriptor backends
~~~~~~~~~~~~~~~~~~~
The descriptor backend (which contains URI, mediatype, charset, language and
other resource-header data) is by default stored in a flat index DB. No
additional backends available at this time.

TODO: a file-based header storage or perhaps even an Apache mod_asis
compatible storage are under consideration. Depending on query/maintenance
requirements.

Rule Syntax
~~~~~~~~~~~
rules.drop::

  # hostpath
  [^/]*zedo\.com.*

Matching DROP rules deny access to the origin server, and instead serve a HTML
or image placeholder.

rules.nocache::

  # hostpath            
  [^/]*gmail\.com.*

A matching NOCACHE rule bypasses the caching for a request, serving directly 
from the origin server or the next proxy on the line.

Both DROP and NOCACHE rule-format will change to include matching on protocol.

rules.sort::

  # proto  hostpath               replacement             root
  *        (.*)                   
  *        [^/]*youtube\.com.*    /my/dir/youtube/\1.flv  mydir/

SORT rules currently prefix the cache-location with a tag, in above example the
location under ROOT will be ``mydir/``. If the ``--archive`` option is in effect
it is prefixed to this tag. Note that ``--nodir`` is applied after prefixing.

This feature is under development.
Rewriting content based on above message matching is planned.



CHANGELOG.rst
TODO htcache version 1.0 (planned)
  * Tested FTP use.
  * RFC 2616 adherence.
  * On-line cache maintenance.  
  * Static browsing of cached resources.  
  * Multiple mechanisms to reduce content duplication.
  * Filter content using URL regex matching.  
  * Keep favicon until user approves changes, set custom favicons.
  * In-browser JS control served with (x)HTML resources to allow 
    in-navigation proxy operations on the current resource.
  * Keep version and navigation history for certain resources.

2012-03-14 htcache version 0.9 
2012-03-14 htcache version 0.8 
2012-03-14 htcache version 0.7 
2012-03-14 htcache version 0.6 
2012-03-14 htcache version 0.5 
2012-03-14 htcache version 0.4 (development)
  * All this time forgot to bump version number.
  * Capture: Re-introduce content hashing but only for specific resources.
    Track history by these checksums, and also track X-Relationship (nav)
    relations between them.
  * Proc: implement idea to defer processing to external script.
    Makes for a very dirty HTTP proxy triggered workflow. 

2010-09-28 htcache version 0.3 (mpe)
  * Many changes and a bit of history corruption in the branches, I made a mess
    of it but as of now most tests pass again. Development went on in too many
    disparate directions and I didn't handle that well.
  * However some new tests were created and most of the tests are passing.
    FTP support has not returned yet.
  * NoCache and Drop are working as before (cache bypass and request blocker)
  * Join has been added to rewrite URL's (effectively joining them if possible with
    existing downloads). This works to reduce duplication for known URLs.
  * Mapping known headers to proper case.

2010-09-28 htcache version 0.2 (mpe)
  * rename to htcache
  * added support for alternative cache directory layouts using modular 
    cache backends  
  * added descriptor cache for resource metadata  
  * allow simple queries on resource metadata  
  * added support to block request based on URL matching
  * added fancy html and image placeholder for blocked requests  
  * added support to bypass cache based on URL matching
  * created Sys-V init sh script
  * including HTTP Via header
  * Updated Makefile for tar'ed dists and incremental snapshots.
  * various more changes from the past two years, see revision log

2008-31-01 http-replicator version 4.0alpha2
  * added GPL licence file
  * generalized fiber.py by using generator name instead of hardcoded string
  * removing partial file in cache after 403 forbidden
  * flushing every line in debug mode
  * fix: no longer setting mtime to -1 if server does not send Last-Modified
  * fix: handling empty command lines correctly

2008-01-01 http-replicator version 4.0alpha1
  * rewrite from scratch
  * replaced restrictive asyncore scheduler with new 'fiber' system
  * new feature: server-side download resuming
  * new feature: fpt support
  * new feature: bandwidth shaping
  * new feature: ipv6 support
  * new feature: frozen transactions killed after configurable timeout
  * new feature: rudimentary off-line mode
  * fixed race condition that prevented joining of simultaneously started downloads
  * currently missing feature: cache browsing

2004-11-27 http-replicator version 3.0
  * new feature: cache brower on proxy address
  * new feature: client-side support for partial content
  * added alias option for caching mirrors on same location
  * added check to prevent access outside of cache through symlinks
  * added header length restriction to fight infinite request server attacks
  * created man pages for http-replicator and http-replicator_maintenance
  * fixed timestamp bug; files are now properly closed before changing mtime
  * suppressed size warning for chunked data

2004-08-15 http-replicator version 2.1
  * integrated daemon code in http-replicator
  * changed init.d and cron script to bash
  * moved settings from configuration file to /etc/default/http-replicator
  * introduced optparse module for command line parsing
  * introduced logging module for output
  * added support for an external proxy server
  * added support for an external proxy requiring authentication

2004-05-01 http-replicator version 2.0
  * added support for HTTP/1.1
  * replicator is now suitable for maintaining a gentoo package cache
  * fixed problem with absolute urls
  * added posting support
  * added support for servers that use LF in header instead of CRLF
  * added a command line system
  * fixed security issues
  * improved traceback message for unhandled exceptions
  * fixed problem with incomplete files after a server crash
  * fixed problems with select
  * fixed size calculation in cron script

2004-02-06 http-replicator version 1.0
  * initial release.



Params.py
import os, re, socket, sys
# XXX: Dont use cjson, its buggy, see comments at
# http://pypi.python.org/pypi/python-cjson
# use jsonlib or simplejson
try:
    import simplejson as _json
except:
    import json as _json

json_read = _json.loads
json_write = _json.dumps


## Main: determine runtime config from constants and ARGV

_args = iter( sys.argv )

VERSION = 0.4

# proxy params
PROG = _args.next()
PORT = 8080
ROOT = os.getcwd() + os.sep
PID_FILE = '/var/run/htcache.pid'
VERBOSE = 0
TIMEOUT = 15
FAMILY = socket.AF_INET
STATIC = False
ONLINE = True # XXX:bvb: useless..
LIMIT = False # XXX unused

# misc. program params
LOG = False
DEBUG = False
MODE = [] # emtpy for normal operation, function list for maintenance

# proxy rule files
DROP = []
DROP_FILE = '/etc/htcache/rules.drop'
JOIN = []
JOIN_FILE = '/etc/htcache/rules.join'
NOCACHE = []
NOCACHE_FILE = '/etc/htcache/rules.nocache'
REWRITE = []
REWRITE_FILE = '/etc/htcache/rules.rewrite'

# cache backend
CACHE = 'caches.FileTree'
ARCHIVE = ''
ENCODE_PATHSEP = ''
FileTreeQ_SORT = True
FileTreeQ_ENCODE = False
# non user-configurable
MAX_PATH_LENGTH = 256
MAXCHUNK = 1448 # maximum lan packet?
TIMEFMT = '%a, %d %b %Y %H:%M:%S GMT'
ALTTIMEFMT = '%a, %d %b %H:%M:%S CEST %Y' # foksuk.nl
PARTIAL = '.incomplete'
IMG_TYPE_EXT = 'png','jpg','gif','jpeg','jpe'
DATA_DIR = '/var/lib/htcache/'
RESOURCES = DATA_DIR+'resource.db'
HTML_PLACEHOLDER = DATA_DIR+'filtered-placeholder.html'
IMG_PLACEHOLDER = DATA_DIR+'forbidden-sign.png'
PROXY_INJECT = True
PROXY_INJECT_JS = DATA_DIR+'dhtml.js'
PROXY_INJECT_CSS = DATA_DIR+'dhtml.css'

# Static mode, query params
CMD = None
CMD_ARGS = []

PRUNE = False
MAX_SIZE_PRUNE = 11*(1024**2)
INTERACTIVE = False

USAGE = '''usage: %(PROG)s [options]

  -h --help          show this help message and exit

Proxy:
  -p --port PORT     listen on this port for incoming connections, default %(PORT)i
  -r --root DIR      set cache root directory, default current: %(ROOT)s
     --static        static mode; assume files never change
     --offline       offline mode; never connect to server
     --limit RATE    FIXME: limit download rate at a fixed K/s
     --daemon LOG    daemonize process and print PID, route output to LOG
     --debug         switch from gather to debug output module

Cache:
  -f RESOURCES

Rules:
     --drop FILE     filter requests for URI's based on regex patterns.
                     read line for line from file, default %(DROP_FILE)s.
     --nocache FILE  TODO: bypass caching for requests based on regex pattern.
     --rewrite FILE  Filter any webresource by selecting on URL or 

Query
     --media-image

Misc.:
     --check-refs    TODO: iterate cache references.
     --check-sortlist 
                     TODO: iterate cache references, 
  -t --timeout SEC   break connection after so many seconds of inactivity,
                     default %(TIMEOUT)i
  -6 --ipv6          try ipv6 addresses if available
  -v --verbose       increase output, use twice to show http headers
                    ''' % locals()


for _arg in _args:

    if _arg in ( '-h', '--help' ):
        sys.exit( USAGE )

    elif _arg in ( '-p', '--port' ):
        try:
            PORT = int( _args.next() )
            assert PORT > 0
        except:
            sys.exit( 'Error: %s requires a positive numerical argument' % _arg )

    elif _arg in ( '--drop', ):
        try:
            DROP_FILE = os.path.realpath(_args.next())
            assert os.path.exists(DROP_FILE)
        except:
            sys.exit( 'Error: %s requires an filename argument' % _arg )
    elif _arg in ( '--nocache', ):
        try:
            NOCACHE_FILE = os.path.realpath(_args.next())
            assert os.path.exists(NOCACHE_FILE)
        except:
            sys.exit( 'Error: %s requires an filename argument' % _arg )
    elif _arg in ( '--rewrite', ):
        try:
            REWRITE_FILE = os.path.realpath(_args.next())
            assert os.path.exists(REWRITE_FILE)
        except:
            sys.exit( 'Error: %s requires an filename argument' % _arg )

    elif _arg in ( '-r', '--root' ):
        try:
            ROOT = os.path.realpath( _args.next() ) + os.sep
            assert os.path.isdir( ROOT )
        except StopIteration:
            sys.exit( 'Error: %s requires a directory argument' % _arg )
        except:
            sys.exit( 'Error: invalid cache directory %s' % ROOT )
    elif _arg in ( '-v', '--verbose' ):
        VERBOSE += 1
    elif _arg in ( '--nodir' ):
        pass # XXX
        #try:
        _args.next()
        #except:
        #    sys.exit( 'Error: %s requires argument' % _arg )
    elif _arg in ( '--cache' ):
        try:
            CACHE = _args.next()
        except:
            sys.exit( 'Error: %s requires argument' % _arg )
    elif _arg in ( '-t', '--timeout' ):
        try:
            TIMEOUT = int( _args.next() )
            assert TIMEOUT > 0
        except:
            sys.exit( 'Error: %s requires a positive numerical argument' % _arg )
    elif _arg in ( '-6', '--ipv6' ):
        FAMILY = socket.AF_UNSPEC
    elif _arg == '--static':
        STATIC = True
    elif _arg == '--daemon':
        LOG = _args.next()
    elif _arg == '--debug':
        DEBUG = True
    elif _arg in ('--pid-file',):
        PID_FILE = _args.next()

    elif _arg in ('-f','--resource'):
        RESOURCES = _args.next()
    elif _arg in ('--prune',):
        PRUNE = True

    elif _arg in ('--print-allrecords',):
        CMD='print-all-records'
    elif _arg in ('--print-record',):
        CMD='print-record'
        CMD_ARGS=_args.next()
    elif _arg in ('--find-records',):
        CMD='find-records'
        CMD_ARGS=_args.next()

#    elif _arg in ('--check-joinlist',):
#        MODE.append(check_joinlist)
    elif _arg in ('--run-join-rules',):
        MODE.append('run-join')

    elif _arg in ('--check-cache',):
        CHECK = 'cache'
#    elif _arg in ('--validate-cache',):
#        CHECK = 'validate'
    elif _arg in ('--check-files',):
        CHECK = 'files'

    else:
        sys.exit( 'Error: invalid option %r' % _arg )


def log(msg, threshold=0):
    """
    Not much of a log..
    Output if VERBOSE >= threshold
    """
    #assert not threshold == 0
    # see fiber.py which manages stdio
    if VERBOSE >= threshold:
        print msg

def parse_droplist(fpath=DROP_FILE):
    global DROP
    DROP = []
    DROP.extend([(p.strip(), re.compile(p.strip())) for p in
        open(fpath).readlines() if p.strip() and not p.startswith('#')])

def parse_nocache(fpath=NOCACHE_FILE):
    global NOCACHE
    NOCACHE = []
    NOCACHE.extend([(p.strip(), re.compile(p.strip())) for p in
        open(fpath).readlines() if p.strip() and not p.startswith('#')])


def parse_joinlist(fpath=JOIN_FILE):
    global JOIN
    JOIN = []
    if os.path.isfile(fpath):
        JOIN.extend([(p.strip(), re.compile('^'+p.split(' ')[0].strip()+'$')) for p in
            open(fpath).readlines() if p.strip() and not p.strip().startswith('#')])

def parse_rewritelist(fpath=REWRITE_FILE):
    global REWRITE
    REWRITE = []
    for p in open(fpath).readlines():
        if not p.strip() or p.strip().startswith('#'):
            continue
        match, replace = p.strip().split('\t')
        REWRITE.append((re.compile(match), replace))

# XXX: first need to see working
def parse_rewritelist_(fpath=REWRITE_FILE):
    global REWRITE
    REWRITE_RULES = []
    REWRITE = {}
    for p in open(fpath).readlines():
        if not p.strip() or p.strip().startswith('#'):
            continue
            
        # Parse line and cleanup, compile rule
        fields = p.strip().split('\t')
        patterns = [ re.compile(f) if f != '.*' else None for f in fields[:-1] ]

        mime_pattern, hostinfo_pattern, path_pattern, entity_match = patterns
        entity_replace = fields[-1]
      
        # Get rule number
        if entity_replace in REWRITE_RULES:
            idx = REWRITE_RULES.index(entity_replace)
        else:
            idx = len(REWRITE_RULES)
            REWRITE_RULES.append(entity_replace)

        # Store new content rewrite rules
#        REWRITE[] = 
        REWRITE.append((
                mime_pattern,
                hostinfo_pattern,
                path_pattern,
                entity_mathc,
                entity_replace
            ))

def match_rewrite(mediatype, hostinfo, path):
    pass
#/XXX

def validate_joinlist(fpath=JOIN_FILE):
    """
    Read all double commented lines as URLs, use on next first pattern line.
    """
    lines = [path[2:].strip() for path in open(fpath).readlines() if path.strip() and path.strip()[1]=='#']
    for path in lines:
        match = False
        for line, regex in JOIN:
            m = regex.match(path)
            if m:
                #print 'Match', path, m.groups()
                match = True
        if not match:
            print "Error: no match for", path

def format_info():
    """
    Return JSON for config.
    """
    return json_write({
        "htcache": { 
            "runtime": {
                "program": PROG,
            },
            "config": {
                "port": PORT,
                "root": ROOT,
                "pid-file": PID_FILE,
                "verboseness": VERBOSE,
                "timeout": TIMEOUT,
                "socket-family": FAMILY,
                "cache-type": CACHE,
                "join-file": JOIN_FILE,
                "drop-file": DROP_FILE,
                "nocache-file": NOCACHE_FILE,
                "rewrite-file": REWRITE_FILE,
            },
            "statistics": {
                "rules": {
                        "drop": len(DROP),
                        "join": len(JOIN),
                        "nocache": len(NOCACHE),
                        "rewrite": len(REWRITE),
                    }
                }
            }
        })

descriptor_storage_type = None

if __name__ == '__main__':
    #parse_joinlist()
    #validate_joinlist()
    parse_rewritelist()

